node/url.ts

```diff
@@ -116,6 +175,296 @@ export function format(
+      return `/${self.hostname.slice(i)}${rest}`;
+    }
+  }
+  return rest;
+}
+
+/**
+ * The url.parse() method takes a URL string, parses it, and returns a URL object..
+ *
+ * @param url The URL string to parse.
+ * @param parseQueryString If `true`, the query property will always be set to an object returned by the querystring module's parse() method. If false,
+ * the query property on the returned URL object will be an unparsed, undecoded string. Default: false.
+ * @param slashesDenoteHost If `true`, the first token after the literal string // and preceding the next / will be interpreted as the host
+ */
+export function parse(
+  url: string,
+  parseQueryString: boolean,
+  slashesDenoteHost: boolean,
+) {
+  // Copy chrome, IE, opera backslash-handling behavior.
+  // Back slashes before the query string get converted to forward slashes
+  // See: https://code.google.com/p/chromium/issues/detail?id=25916
+  let hasHash = false;
+  let start = -1;
+  let end = -1;
+  let rest = "";
+  let lastPos = 0;
+  for (let i = 0, inWs = false, split = false; i < url.length; ++i) {
+    const code = url.charCodeAt(i);
+
+    // Find first and last non-whitespace characters for trimming
+    const isWs = code === CHAR_SPACE ||
+      code === CHAR_TAB ||
+      code === CHAR_CARRIAGE_RETURN ||
+      code === CHAR_LINE_FEED ||
+      code === CHAR_FORM_FEED ||
+      code === CHAR_NO_BREAK_SPACE ||
+      code === CHAR_ZERO_WIDTH_NOBREAK_SPACE;
+    if (start === -1) {
+      if (isWs) continue;
+      lastPos = start = i;
+    } else if (inWs) {
+      if (!isWs) {
+        end = -1;
+        inWs = false;
+      }
+    } else if (isWs) {
+      end = i;
+      inWs = true;
+    }
+
+    // Only convert backslashes while we haven't seen a split character
+    if (!split) {
+      switch (code) {
+        case CHAR_HASH:
+          hasHash = true;
+        // Fall through
+        case CHAR_QUESTION_MARK:
+          split = true;
+          break;
+        case CHAR_BACKWARD_SLASH:
+          if (i - lastPos > 0) rest += url.slice(lastPos, i);
+          rest += "/";
+          lastPos = i + 1;
+          break;
+      }
+    } else if (!hasHash && code === CHAR_HASH) {
+      hasHash = true;
+    }
+  }
+
+  // Check if string was non-empty (including strings with only whitespace)
+  if (start !== -1) {
+    if (lastPos === start) {
+      // We didn't convert any backslashes
+
+      if (end === -1) {
+        if (start === 0) rest = url;
+        else rest = url.slice(start);
+      } else {
+        rest = url.slice(start, end);
+      }
+    } else if (end === -1 && lastPos < url.length) {
+      // We converted some backslashes and have only part of the entire string
+      rest += url.slice(lastPos);
+    } else if (end !== -1 && lastPos < end) {
+      // We converted some backslashes and have only part of the entire string
+      rest += url.slice(lastPos, end);
+    }
+  }
+
+  const self = new URL(url);
```

Comment:
I think `url.parse` should return url.Url object, not a global URL object

You can check it in Node.js repl like the below
```
> url.parse("https://example.com") instanceof url.Url
true
> url.parse("https://example.com") instanceof URL
false
```